src/core/lib/promise/activity.h

```diff
@@ -51,108 +51,93 @@ class Wakeable {
+    std::swap(wakeable_and_arg_, other.wakeable_and_arg_);
     return *this;
   }
 
   // Wake the underlying activity.
-  void Wakeup() { Take()->Wakeup(); }
+  void Wakeup() { Take().Wakeup(); }
 
   template <typename H>
   friend H AbslHashValue(H h, const Waker& w) {
-    return H::combine(std::move(h), w.wakeable_);
+    return H::combine(H::combine(std::move(h), w.wakeable_and_arg_.wakeable),
+                      w.wakeable_and_arg_.arg);
   }
 
   bool operator==(const Waker& other) const noexcept {
-    return wakeable_ == other.wakeable_;
+    return wakeable_and_arg_ == other.wakeable_and_arg_;
   }
 
   bool operator!=(const Waker& other) const noexcept {
-    return wakeable_ != other.wakeable_;
+    return !operator==(other);
   }
 
   std::string ActivityDebugTag() {
-    return wakeable_ == nullptr ? "<unknown>" : wakeable_->ActivityDebugTag();
+    return wakeable_and_arg_.ActivityDebugTag();
   }
 
- private:
-  friend class AtomicWaker;
-
-  Wakeable* Take() {
-    return std::exchange(wakeable_, promise_detail::unwakeable());
-  }
-
-  Wakeable* wakeable_;
-};
-
-// An atomic variant of Waker - this type is non-copyable and non-movable.
-class AtomicWaker {
- public:
-  explicit AtomicWaker(Wakeable* wakeable) : wakeable_(wakeable) {}
-  AtomicWaker() : AtomicWaker(promise_detail::unwakeable()) {}
-  explicit AtomicWaker(Waker waker) : AtomicWaker(waker.Take()) {}
-  ~AtomicWaker() { wakeable_.load(std::memory_order_acquire)->Drop(); }
-  AtomicWaker(const AtomicWaker&) = delete;
-  AtomicWaker& operator=(const AtomicWaker&) = delete;
-  AtomicWaker(AtomicWaker&& other) noexcept = delete;
-  AtomicWaker& operator=(AtomicWaker&& other) noexcept = delete;
-
-  // Wake the underlying activity.
-  void Wakeup() { Take()->Wakeup(); }
-
-  // Return true if there is a not-unwakeable wakeable present.
-  bool Armed() const noexcept {
-    return wakeable_.load(std::memory_order_relaxed) !=
-           promise_detail::unwakeable();
-  }
-
-  // Set to some new waker
-  void Set(Waker waker) {
-    wakeable_.exchange(waker.Take(), std::memory_order_acq_rel)->Wakeup();
+  // This is for tests to assert that a waker is occupied or not.
+  bool is_unwakeable() const {
+    return wakeable_and_arg_.wakeable == promise_detail::unwakeable();
   }
 
  private:
-  Wakeable* Take() {
-    return wakeable_.exchange(promise_detail::unwakeable(),
-                              std::memory_order_acq_rel);
+  struct WakeableAndArg {
+    Wakeable* wakeable;
+    void* arg;
+
+    void Wakeup() { wakeable->Wakeup(arg); }
+    void Drop() { wakeable->Drop(arg); }
+    std::string ActivityDebugTag() const {
+      return wakeable == nullptr ? "<unknown>"
+                                 : wakeable->ActivityDebugTag(arg);
+    }
+    bool operator==(const WakeableAndArg& other) const noexcept {
+      return wakeable == other.wakeable && arg == other.arg;
+    }
+  };
+
+  WakeableAndArg Take() {
+    return std::exchange(wakeable_and_arg_,
```

Comment:
On some platforms, this will be a mutex-protected exchange given that not all of them have 128-bit atomics. Given that you're exchanging the `wakeable` really, can we just keep two pointers (one atomic) and only exchange the `wakeable`?  AFAICT, keeping the `arg` pointer is harmless.

Reply:
This is std::exchange - it's not atomic at all.

Reply:
it's shorthand for:

```
auto r = wakeable_and_arg_;
wakeable_and_arg_ = {promise_detail::unwakeable(), nullptr};
return r;
```

we use the pattern relatively frequently

Reply:
Oh sorry, I completely misread it. Was looking at HEAD and though this is atomic too.

Reply:
Yeah... I just nuked the AtomicWaker stuff - seemed like a good idea at the time, but we don't use it: you always need to update some state and assign the waker atomically so you need some higher level synchronization primitive.

Reply:
That's a good idea. Thanks!